$NetBSD$

Caution: the extra cpp checking (the three hunks containing "Should be
one line") is manually hacked and not reflected in configure.in; do
not regenerate configure with autoconf without merging it. This
checking is required because somewhere between gcc 4.1 and 4.5 cpp
started doing regrettable things with backslash-escaped newlines.

* Fix build on NetBSD 6.99.23, gcc -E -traditional-cpp with native X
  results syntax error. Use -traditional-cpp for Makefile generation only.

* Give x86_64 a machine of its own when the operating system is not one of
  the two spelled out above.  configure already learned `opsys=dragonfly'
  from the hunk further down, but nothing set `machine', so DragonFly stopped
  at

	configure: error: Emacs hasn't been ported to `x86_64-pc-dragonfly'
	systems.

  editors/emacs20 carries the same arm; this one was left behind.  The netbsd
  and openbsd cases are earlier in the same switch, so they still win.

--- configure.orig	2003-03-18 09:19:12.000000000 -0500
+++ configure
@@ -821,10 +821,18 @@
       powerpc-*-netbsd*) machine=macppc ;;
       sparc*-*-netbsd*)	machine=sparc ;;
       vax-*-netbsd*)	machine=vax ;;
-      arm-*-netbsd*)	machine=arm ;;
+      arm*-*-netbsd*)	machine=arm ;;
+      x86_64-*-netbsd*)	machine=amd64 ;;
+      hppa-*-netbsd*)	machine=hp800 ;;
+      shle-*-netbsd*)	machine=sh3el ;;
     esac
   ;;
 
+  ## Darwin / Mac OS X
+  powerpc-apple-darwin* )
+    machine=powermac opsys=darwin ;;
+  i386-apple-darwin* )
+    machine=intel386 opsys=darwin ;;
   ## OpenBSD ports
   *-*-openbsd* )
     opsys=openbsd
@@ -909,6 +917,20 @@
     machine=macppc opsys=gnu-linux
   ;;
 
+  x86_64-*-linux* )
+    machine=amd64 opsys=gnu-linux
+  ;;
+
+  ## AMD/Intel x86_64 machines where we don't care about the manufacturer
+  x86_64-*-* )
+    machine=amd64
+    case "${canonical}" in
+      *-freebsd* )		opsys=freebsd ;;
+      *-dragonfly* )		opsys=dragonfly ;;
+      ## Otherwise, we'll fall through to the generic opsys code at the bottom.
+    esac
+  ;;
+
   ## Altos 3068
   m68*-altos-sysv* )
     machine=altos opsys=usg5-2
@@ -1659,6 +1681,7 @@
       *-sysv4.2uw* )	  	opsys=unixware; NON_GNU_CPP=/lib/cpp ;;
       *-386bsd* )	        opsys=386bsd ;;
       *-freebsd* )	        opsys=freebsd ;;
+      *-dragonfly* )	        opsys=dragonfly ;;
       *-nextstep* )             opsys=nextstep ;;
       ## Otherwise, we'll fall through to the generic opsys code at the bottom.
     esac
@@ -2052,15 +2075,20 @@
   CPP="${CC-cc} -E"
   # On the NeXT, cc -E runs the code through the compiler's parser,
   # not just through cpp.
+  # And, gcc 4.5? and up does not preserve escaped newlines, which is
+  # fatal for preprocessing makefiles.
   cat > conftest.$ac_ext <<EOF
 #line 2057 "configure"
 #include "confdefs.h"
 #include <assert.h>
 Syntax Error
+Should be one line.
 EOF
-ac_try="$ac_cpp conftest.$ac_ext >/dev/null 2>conftest.out"
+ac_try="$ac_cpp conftest.$ac_ext >conftest2.out 2>conftest.out"
 { (eval echo configure:2063: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
-ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"`
+ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"; \
+    grep -L "Should  *be one line" conftest2.out \
+    `
 if test -z "$ac_err"; then
   :
 else
@@ -2068,16 +2096,19 @@
   echo "configure: failed program was:" >&5
   cat conftest.$ac_ext >&5
   rm -rf conftest*
-  CPP="${CC-cc} -E -traditional-cpp"
+  CPP="${CC-cc} -E"
   cat > conftest.$ac_ext <<EOF
 #line 2074 "configure"
 #include "confdefs.h"
 #include <assert.h>
 Syntax Error
+Should be one line.
 EOF
-ac_try="$ac_cpp conftest.$ac_ext >/dev/null 2>conftest.out"
+ac_try="$ac_cpp conftest.$ac_ext >conftest2.out 2>conftest.out"
 { (eval echo configure:2080: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
-ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"`
+ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"; \
+    grep -L "Should  *be one line" conftest2.out \
+    `
 if test -z "$ac_err"; then
   :
 else
@@ -2091,10 +2122,13 @@
 #include "confdefs.h"
 #include <assert.h>
 Syntax Error
+Should be one line.
 EOF
-ac_try="$ac_cpp conftest.$ac_ext >/dev/null 2>conftest.out"
+ac_try="$ac_cpp conftest.$ac_ext >conftest2.out 2>conftest.out"
 { (eval echo configure:2097: \"$ac_try\") 1>&5; (eval $ac_try) 2>&5; }
-ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"`
+ac_err=`grep -v '^ *+' conftest.out | grep -v "^conftest.${ac_ext}\$"; \
+    grep -L "Should  *be one line" conftest2.out \
+    `
 if test -z "$ac_err"; then
   :
 else
@@ -2416,7 +2450,7 @@
 if test x"${OVERRIDE_CPPFLAGS}" != x; then
   CPPFLAGS="${OVERRIDE_CPPFLAGS}"
 else
-  CPPFLAGS="$c_switch_system $c_switch_machine $CPPFLAGS"
+  CPPFLAGS="`echo $c_switch_system $c_switch_machine $CPPFLAGS | sed 's,-L[^ ]*,,g'`"
 fi
 
 
@@ -5583,24 +5617,24 @@
   echo "$ac_t""yes" 1>&6
   # EGifPutExtensionLast only exists from version libungif-4.1.0b1.
 # Earlier versions can crash Emacs.
-      echo $ac_n "checking for EGifPutExtensionLast in -lungif""... $ac_c" 1>&6
-echo "configure:5588: checking for EGifPutExtensionLast in -lungif" >&5
-ac_lib_var=`echo ungif'_'EGifPutExtensionLast | sed 'y%./+-%__p_%'`
+      echo $ac_n "checking for EGifPutExtensionTrailer in -lgif""... $ac_c" 1>&6
+echo "configure:5588: checking for EGifPutExtensionTrailer in -lgif" >&5
+ac_lib_var=`echo ungif'_'EGifPutExtensionTrailer | sed 'y%./+-%__p_%'`
 if eval "test \"`echo '$''{'ac_cv_lib_$ac_lib_var'+set}'`\" = set"; then
   echo $ac_n "(cached) $ac_c" 1>&6
 else
   ac_save_LIBS="$LIBS"
-LIBS="-lungif  $LIBS"
+LIBS="-lgif  $LIBS"
 cat > conftest.$ac_ext <<EOF
 #line 5596 "configure"
 #include "confdefs.h"
 /* Override any gcc2 internal prototype to avoid an error.  */
 /* We use char because int might match the return type of a gcc2
     builtin and then its argument prototype would still apply.  */
-char EGifPutExtensionLast();
+char EGifPutExtensionTrailer();
 
 int main() {
-EGifPutExtensionLast()
+EGifPutExtensionTrailer()
 ; return 0; }
 EOF
 if { (eval echo configure:5607: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext}; then
@@ -9120,9 +9154,63 @@
   cat >> confdefs.h <<\EOF
 #define size_t unsigned
 EOF
+
+fi
+
+
+# NETBSD: NetBSD's newer run-time linker fix.
+if test $opsys = netbsd; then
+  if test -f /usr/lib/crti.o; then
+    cat >> confdefs.h <<\EOF
+#define HAVE_NETBSD_CRTI 1
+EOF
+
+  fi
+  if test -f /usr/lib/crtn.o; then
+    cat >> confdefs.h <<\EOF
+#define HAVE_NETBSD_CRTN 1
+EOF
+
+  fi
+fi
+
+# XXX: handcoded test for ptyfs
+echo $ac_n "checking for ptyfs""... $ac_c" 1>&6
+echo "configure:9095: checking for ptyfs" >&5
+cat > conftest.$ac_ext <<EOF
+#include "confdefs.h"
+#include <stdlib.h>
+#include <fcntl.h>
+
+int
+main(void)
+{
+	int fd;
 
+	fd = posix_openpt(O_RDWR);
+	if (fd == -1)
+		return 1;
+	close(fd);
+	return 0;
+}
+EOF
+if { (eval echo configure:9185: \"$ac_link\") 1>&5; (eval $ac_link) 2>&5; } && test -s conftest${ac_exeext} && (./conftest; exit) 2>/dev/null
+then
+  ac_cv_have_ptyfs=yes
+else
+  echo "configure: failed program was:" >&5
+  cat conftest.$ac_ext >&5
+  rm -fr conftest*
+  ac_cv_have_ptyfs=no
 fi
+rm -fr conftest*
 
+echo "$ac_t""$ac_cv_have_ptyfs" 1>&6
+if test $ac_cv_have_ptyfs = yes; then
+  cat >> confdefs.h <<\EOF
+#define HAVE_PTYFS 1
+EOF
+fi
 
 # Set up the CFLAGS for real compilation, so we can substitute it.
 CFLAGS="$REAL_CFLAGS"
@@ -9392,6 +9480,8 @@
 do
   case "\$ac_option" in
   -recheck | --recheck | --rechec | --reche | --rech | --rec | --re | --r)
+	: Avoid regenerating within pkgsrc
+	exit 0
     echo "running \${CONFIG_SHELL-/bin/sh} $0 $ac_configure_args --no-create --no-recursion"
     exec \${CONFIG_SHELL-/bin/sh} $0 $ac_configure_args --no-create --no-recursion ;;
   -version | --version | --versio | --versi | --vers | --ver | --ve | --v)
@@ -9735,7 +9825,7 @@
   sed -e '1,/start of cpp stuff/d'\
       -e 's,/\*\*/#\(.*\)$,/* \1 */,' \
       < Makefile.c > junk.c
-  $CPP $undefs -I. -I$top_srcdir/src $CPPFLAGS junk.c | \
+  $TRADCPP $undefs -I. -I$top_srcdir/src $CPPFLAGS junk.c | \
       sed -e 's/^ /	/' -e '/^#/d' -e '/^[ 	]*$/d' > junk2.c
   cat junk1.c junk2.c > Makefile.new
   rm -f junk.c junk1.c junk2.c
@@ -9751,7 +9841,7 @@
   sed -e '1,/start of cpp stuff/d'\
       -e 's,/\*\*/#\(.*\)$,/* \1 */,' \
       < Makefile.c > junk.c
-  $CPP $undefs -I. -I$top_srcdir/src $CPPFLAGS junk.c | \
+  $TRADCPP $undefs -I. -I$top_srcdir/src $CPPFLAGS junk.c | \
       sed -e 's/^ /	/' -e '/^#/d' -e '/^[ 	]*$/d' > junk2.c
   cat junk1.c junk2.c > Makefile.new
   rm -f junk.c junk1.c junk2.c
